File: tests/ej32/ESP32.java
package ej32;

public class ESP32 {
    // constants
    public static final int LED_BUILTIN  = 2;
    public static final int INPUT        = 0;
    public static final int OUTPUT       = 1;
    public static final int INPUT_PULLUP = 2;
    public static final int LOW          = 0;
    public static final int HIGH         = 1;
    // methods
    public static void pinMode(int pin, int mode)    {}
    public static int  digitalRead(int pin)          { return 0; }
    public static void digitalWrite(int pin, int v)  {}
    public static int  analogRead(int pin)           { return 0; }
    public static void analogWrite(int pin, int v)   {}
    public static void ledcAttachPin(int pin, int v) {}
    public static void ledcSetup(int ch, int freq, int v) {}
    public static void ledcWriteTone(int ch, int v)  {}
}

File: src/esp32.cpp
#include "ucode.h"

#define CODE(s, g)  { s, [](Thread &t){ g; }, ACL_BUILTIN }
#define POP         t.pop()
#define PUSH(v)     t.push(v)
#if ESP32
#define analogWrite(c,v,mx) ledcWrite((c),(8191/mx)*min((int)(v),mx))
#endif // ESP32
static Method _esp32[] = {
    ///
    /// @definegroup ESP32 supporting functions
    /// @{
#if ARDUINO && ESP32
    CODE("pinMode",      DU m = POP; pinMode(POP, m)),
    CODE("digitalRead",  PUSH(digitalRead(POP))),
    CODE("digitalWrite", DU v = POP; digitalWrite(POP, v)),
    CODE("analogRead",   PUSH(analogRead(POP))),
    CODE("analogWrite",  DU v = POP; analogWrite(POP, v, 255)),
    CODE("ledcAttachPin",DU v = POP; ledcAttachPin(POP, v)),
    CODE("ledcSetup",    DU v = POP; DU freq=POP; ledcSetup(POP, freq, v)),
    CODE("ledcWriteTone",DU v = POP; ledcWriteTone(POP, v)),
#else
    /// for debugging
    CODE("pinMode",      DU m = POP; DU p = POP),
    CODE("digitalRead",  DU p = POP; PUSH(t.IP)),
    CODE("digitalWrite", DU v = POP; DU p = POP),
    CODE("analogRead",   DU p = POP; PUSH(t.IP)),
    CODE("analogWrite",  DU v = POP; DU p = POP),
    CODE("ledcAttachPin",DU v = POP; DU p = POP),
    CODE("ledcSetup",    DU v = POP; DU freq=POP; DU p = POP),
    CODE("ledcWriteTone",DU v = POP; DU p = POP)
#endif // ARDUINO
    /// @}
};
///
/// ESP32 built-in word in ROM, use extern by main program
///
Ucode uESP32(VTSZ(_esp32), _esp32);


